Merge operation/sourceLocation/insertId

scgao opened the pull request:
In this patch, when there are extra subfields in the field, the extra ones will be reserved in the jsonPayload.

scgao committed:
out_stackdriver: added support for operation field

Enable the stackdriver to identify 'operation' field inside the
jsonPayload and add it to the entries.

Testing file path: tests/runtime/out_stackdriver.c

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: added support for operation field

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: added support for operation field

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: update cmake file - add stackdriver_operation.c

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: update cmake file - add stackdriver_operation.c

Signed-off-by: scgao <scgao@google.com>

scgao committed:
tests:runtime:out_stackdriver: unit tests for adding 'operation' field
in stackdriver

Signed-off-by: scgao <scgao@google.com>

scgao committed:
tests:runtime:out_stackdriver: unit tests for adding 'operation' field
in stackdriver

Signed-off-by: scgao <scgao@google.com>

scgao committed:
tests: add out_stackdriver.c

Signed-off-by: scgao <scgao@google.com>

scgao committed:
tests: add out_stackdriver.c

Signed-off-by: scgao <scgao@google.com>

scgao committed:
config: add customized config file for stackdriver

Signed-off-by: scgao <scgao@google.com>

scgao committed:
config: add customized config file for stackdriver

Signed-off-by: scgao <scgao@google.com>

scgao committed:
deal with memory lead

Signed-off-by: scgao <scgao@google.com>

scgao committed:
deal with memory lead

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: remove local config files

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: remove local config files

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: clean up id/producer inside the while loop

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: clean up id/producer inside the while loop

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: update based on the code review
In extrack_operation(), use 2 different keys, one for operation and the
other for logging.googleapis.com/operation;
Deal with memory leak inside extrack_operation;

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: update based on the code review
In extrack_operation(), use 2 different keys, one for operation and the
other for logging.googleapis.com/operation;
Deal with memory leak inside extrack_operation;

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: rename 'operation_existed' as 'operation_extracted

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: rename 'operation_existed' as 'operation_extracted

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: rename 'update_json_payload()' as 'pack_json_payload'

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: rename 'update_json_payload()' as 'pack_json_payload'

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: deal with the heap leak - no heap leak now

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: deal with the heap leak - no heap leak now

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: remove calculate_spec_fields(), calulate the entry size inline

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: remove calculate_spec_fields(), calulate the entry size inline

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: reconstruct the operation_status inside extrack_operation()

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: reconstruct the operation_status inside extrack_operation()

Signed-off-by: scgao <scgao@google.com>

scgao committed:
reconstruct the operation_status inside extrack_operation()

Signed-off-by: scgao <scgao@google.com>

scgao committed:
reconstruct the operation_status inside extrack_operation()

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: construct operation_status to present whether the operation field is valid

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: construct operation_status to present whether the operation field is valid

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: Design several test cases and do runtime test

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: Design several test cases and do runtime test

Signed-off-by: scgao <scgao@google.com>

scgao committed:
test: runtime: unit tests for stackdriver operation

Signed-off-by: scgao <scgao@google.com>

scgao committed:
test: runtime: unit tests for stackdriver operation

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: remove trouble_shooting code

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: remove trouble_shooting code

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: remove trouble_shooting code

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: remove trouble_shooting code

Signed-off-by: scgao <scgao@google.com>

scgao committed:
Update plugins/out_stackdriver/stackdriver.c

Co-authored-by: Yen-Cheng Chou <3055688+StevenYCChou@users.noreply.github.com>

scgao committed:
Update plugins/out_stackdriver/stackdriver.c

Co-authored-by: Yen-Cheng Chou <3055688+StevenYCChou@users.noreply.github.com>
Signed-off-by: scgao <scgao@google.com>

scgao committed:
Update plugins/out_stackdriver/stackdriver.c

Co-authored-by: Yen-Cheng Chou <3055688+StevenYCChou@users.noreply.github.com>
Signed-off-by: scgao <scgao@google.com>

scgao committed:
Update plugins/out_stackdriver/stackdriver.c

Co-authored-by: Yen-Cheng Chou <3055688+StevenYCChou@users.noreply.github.com>

scgao committed:
Update plugins/out_stackdriver/stackdriver.c

Co-authored-by: Yen-Cheng Chou <3055688+StevenYCChou@users.noreply.github.com>
Signed-off-by: scgao <scgao@google.com>

scgao committed:
Update plugins/out_stackdriver/stackdriver.c

Co-authored-by: Yen-Cheng Chou <3055688+StevenYCChou@users.noreply.github.com>

scgao committed:
Update plugins/out_stackdriver/stackdriver.c

Co-authored-by: Yen-Cheng Chou <3055688+StevenYCChou@users.noreply.github.com>
Signed-off-by: scgao <scgao@google.com>

scgao committed:
Update plugins/out_stackdriver/stackdriver_operation.c

Co-authored-by: Yen-Cheng Chou <3055688+StevenYCChou@users.noreply.github.com>
Signed-off-by: scgao <scgao@google.com>

scgao committed:
Update plugins/out_stackdriver/stackdriver_operation.c

Co-authored-by: Yen-Cheng Chou <3055688+StevenYCChou@users.noreply.github.com>

scgao committed:
Update plugins/out_stackdriver/stackdriver_operation.c

Co-authored-by: Yen-Cheng Chou <3055688+StevenYCChou@users.noreply.github.com>
Signed-off-by: scgao <scgao@google.com>

scgao committed:
Update plugins/out_stackdriver/stackdriver_operation.c

Co-authored-by: Yen-Cheng Chou <3055688+StevenYCChou@users.noreply.github.com>

scgao committed:
Update plugins/out_stackdriver/stackdriver_operation.c

Co-authored-by: Yen-Cheng Chou <3055688+StevenYCChou@users.noreply.github.com>
Signed-off-by: scgao <scgao@google.com>

scgao committed:
Update plugins/out_stackdriver/stackdriver_operation.c

Co-authored-by: Yen-Cheng Chou <3055688+StevenYCChou@users.noreply.github.com>

scgao committed:
out_stackdriver: remove operation_status to stackdriver_opration.c and
simplify the return logic of extrack_operation()

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: remove operation_status to stackdriver_opration.c and
simplify the return logic of extrack_operation()

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: simplify the logic of extrack_operation()

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: simplify the logic of extrack_operation()

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: simplify the login in extract_operation

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: simplify the login in extract_operation

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: use flb_sds_copy() to change the value of flb_sds_t

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: use flb_sds_copy() to change the value of flb_sds_t

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: update logic of extract_operation()

scgao committed:
out_stackdriver: update logic of extract_operation()

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: update the logic of extract_operation()

scgao committed:
out_stackdriver: update the logic of extract_operation()

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: simplify the logic of extract_operation()

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: simplify the logic of extract_operation()

scgao committed:
out_stackdriver: simplify the logic of extract_operation()

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: update sub-field of operation field. Make them optional.

Signed-off-by: scgao <scgao@google.com>

scgao committed:
Update plugins/out_stackdriver/stackdriver.c

Co-authored-by: Yen-Cheng Chou <3055688+StevenYCChou@users.noreply.github.com>
Signed-off-by: scgao <scgao@google.com>

scgao committed:
Update plugins/out_stackdriver/stackdriver.c

Co-authored-by: Yen-Cheng Chou <3055688+StevenYCChou@users.noreply.github.com>
Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: Extract all the 'entry_size' calculation and pack into one section

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: simplify the logic of extract_operation

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: extract operation as an object

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: remove useless comment

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: handle empty subfield of 'operation'

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: 'operation' done
Add 'pack_json_payload()' in stackdriver.c to handle multiple special
fields.

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: remove test code

Signed-off-by: scgao <scgao@google.com>

scgao committed:
Merge branch 'update_review_2' of https://github.com/scgao/fluent-bit
Remove the test code so that it will be clear for code review;
Handle the situation that there are extra subfields in 'operation'

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: fix bugs on travis-ci

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: add insertId

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: modify sourceLocation

Signed-off-by: scgao <scgao@google.com>

StevenYCChou commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -461,6 +469,11 @@ static int pack_json_payload(bool operation_extracted, int operation_extra_size,
         msgpack_object_kv* kv = obj->via.map.ptr;
         msgpack_object_kv* const kvend = obj->via.map.ptr + obj->via.map.size;
         for(; kv != kvend; ++kv	) {
+            if (insertId_extracted && strncmp(INSERTID_IN_JSON,kv->key.via.str.ptr, kv->key.via.str.size) == 0 
```
```suggestion
            if (insertId_extracted && strncmp(INSERTID_IN_JSON, kv->key.via.str.ptr, kv->key.via.str.size) == 0 
```

StevenYCChou replied:
I hope we just have a lint/formatter so we don't need to worry about it :)

scgao replied:
Done.

StevenYCChou commented on plugins/out_stackdriver/stackdriver_sourceLocation.c:
```diff
@@ -0,0 +1,123 @@
+                msgpack_object_kv* tmp_p = sub_field.via.map.ptr;
+                msgpack_object_kv* const tmp_pend = sub_field.via.map.ptr + sub_field.via.map.size;
+
+                /* Validate the subfields 
```
I only see the code in this section validating subfields of sourceLocation (or transform from string to numeric). Is the code correct or the comment correct? Please make them consistent. 

If the code is self explained, you don't need the comment.

scgao replied:
Updated.

StevenYCChou commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -624,20 +674,37 @@ static int stackdriver_format(const void *data, size_t bytes,
         flb_sds_destroy(operation_producer);
+        flb_sds_destroy(sourceLocation_file);
+        flb_sds_destroy(sourceLocation_function);
 
```
if you rely on the default value of `sourceLocation_line`, you need to assign the value of `sourceLocation_line` back to `0` each iteration, so it won't use the value extracted from previous iteration.

scgao replied:
Good Point!
This also applies to other variables like operation_first.
I initialize all of them at the beginning of the loop.

scgao committed:
out_stackdriver: updated from the code review
1. remove useless comments
2. initialize some variables at the beginning of the loop inside
stackdriver_format()

Signed-off-by: scgao <scgao@google.com>

scgao committed:
Merge remote-tracking branch 'upstream/master'
Get sync with the upstream master, since unit test framework is updated
Signed-off-by: scgao <scgao@google.com>

scgao committed:
Merge remote-tracking branch 'upstream/master'

scgao committed:
out_stackdriver_test: add unit tests for 'operation'

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: initialize operaion at the begin of loop

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver_test: delete one useless test cases - empty_subfields

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver_test: remove useless testcase

Signed-off-by: scgao <scgao@google.com>

scgao committed:
Merge pull request #3 from scgao/operation_test

Add unit tests for 'operation'

scgao committed:
out_stackdriver: update via comments on the PR
- Do some cosmetic change and align arguments
- Use * at left side of variable
- Define variables at the beginning of the function

Signed-off-by: scgao <scgao@google.com>

scgao committed:
Merge branch 'master' of https://github.com/scgao/fluent-bit

Signed-off-by: scgao <scgao@google.com>

scgao committed:
Merge branch 'master' of https://github.com/scgao/fluent-bit

scgao committed:
Merge branch 'master' of https://github.com/scgao/fluent-bit

scgao committed:
out_stackdriver: replace boolean by FLB_TRUE/FLB_FALSE

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: align arguments

Signed-off-by: scgao <scgao@google.com>

scgao committed:
Merge branch 'master' of https://github.com/scgao/fluent-bit into merge_structured_fields

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: update sourceLocation/insertId
replace boolean by FLB_TRUE/FLB_FALSE
align arguments

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver_test: add unit tests for 'insertId'

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver_test: add unit tests for 'sourceLocation'

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver_test: add unit tests for sourceLocation

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: update insertId
When insertId is invalid, log an error and reject the whole log entry

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver_test: update unit tests for insertId

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: support special field - operation

- Extract 'logging.googleapis.com/operation' field from the jsonPayload
and lift it to the entry-level
- Add unit tests to this feature

Signed-off-by: scgao <scgao@google.com>

koleini committed:
sp: fix memory leak related to groupby string values

Signed-off-by: Masoud Koleini <masoud.koleini@arm.com>

scgao committed:
Merge remote-tracking branch 'upstream/master' into merge_structured_fields

Signed-off-by: scgao <scgao@google.com>
